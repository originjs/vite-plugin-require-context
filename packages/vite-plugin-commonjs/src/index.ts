import { TransformResult } from "esbuild";
import { transformRequire, transformCommonJS, isCommonJS } from "./lib";
import * as fs from "fs";
import { Plugin } from "vite";
import createFilter from "./filter";

export type Options = {
  include?: string | string[] | undefined;
  exclude?: string | string[] | undefined;
  skipPreBuild?: boolean;
  addNamedExports?: boolean;
};

export function viteCommonjs(
  options: Options = { skipPreBuild: false, addNamedExports: false }
): Plugin {
  const filter = createFilter(options.include, options.exclude);
  return {
    name: "originjs:commonjs",
    apply: "serve",
    transform(code: string, id: string): TransformResult {
      if (
        !filter(id) ||
        (options.skipPreBuild && id.indexOf("/node_modules/.vite/") !== -1)
      ) {
        return null;
      }

      let result = transformRequire(code, id);

      if (id.indexOf("/node_modules/.vite/") == -1 && isCommonJS(code)) {
        return transformCommonJS(result.code, options.addNamedExports);
      }

      if (result.replaced) {
        return {
          code: result.code,
          map: null,
          warnings: null,
        };
      }
      return null;
    },
  };
}

export function esbuildCommonjs(include: string[] = []) {
  return {
    name: "originjs:commonjs",
    setup(build) {
      build.onLoad(
        {
          filter: new RegExp("(" + include.join("|") + ").*.js"),
          namespace: "file",
        },
        async ({ path: id }) => {
          const code = fs.readFileSync(id).toString();
          let result = transformRequire(code, id);
          if (result.replaced) {
            return {
              contents: result.code,
              loader: "js",
            };
          }
          return null;
        }
      );
    },
  };
}
