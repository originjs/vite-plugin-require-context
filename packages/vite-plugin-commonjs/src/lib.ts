import { parse as parseCjs } from 'cjs-esm-exports';
import { transformSync } from 'esbuild';

const commonJSRegex: RegExp = /\b(module\.exports|exports\.\w+|exports\s*=\s*|exports\s*\[.*\]\s*=\s*)/;
const requireRegex: RegExp = /(?<!\.)\b_{0,2}require\s*\(\s*(["'`].*?["'`])\s*\)/g;
const IMPORT_STRING_PREFIX: String = "__require_for_vite";
const multilineCommentsRegex = /\/\*(.|[\r\n])*?\*\//gm
const singleCommentsRegex = /([^\:])\/\/.*/g

export interface TransformRequireResult {
  code: string;
  replaced: boolean;
}

export function transformRequire(code: string, id: string): TransformRequireResult {
  let replaced = false;
  // skip if has no require
  if (!/require/.test(code)) {
    return {
      replaced,
      code,
    };
  }
  // empty multiline comments
  code = removeComments(code, multilineCommentsRegex, '/* */');
  // remove singleline comments
  code = removeComments(code, singleCommentsRegex);

  const requireMatches = code.matchAll(requireRegex);
  let importsString = '';
  let packageName = '';
  for (let item of requireMatches) {
    if (!isString(item[1])) {
      console.warn(`Not supported dynamic import, file:${id}`);
      continue;
    }
    replaced = true;
    packageName = `${IMPORT_STRING_PREFIX}_${randomString(6)}`;
    importsString += `import * as ${packageName} from ${item[1].replace(/`/g, `'`)};\n`;
    code = code.replace(item[0], `(${packageName}.default || ${packageName})`);
  }

  if (replaced) {
    code = importsString + code;
  }
  return {
    replaced,
    code,
  };
}

export function isCommonJS(code: string): boolean {
  return commonJSRegex.test(code);
}

export function transformCommonJS(code: string, addNamedExports = false) {
  const transformResult = transformSync(code, { format: "esm" });

  if (!addNamedExports) {
    return transformResult;
  }

  const esmCode = transformResult.code;
  const defaultExportRegex = /export default (.*?);/;
  const defaultExport = esmCode.match(defaultExportRegex)?.[1];

  if (!defaultExport) {
    return transformResult;
  }

  const { exports } = parseCjs('', code);
  if (!exports || !exports.length) {
    return transformResult;
  }

  const namedKey = '__named_exports_for_vite';
  const namedExportsMap: Record<string, string> = exports.reduce(
    (map: Record<string, string>, key: string) => {
      map[key] = `${namedKey}['${key}']`;
      return map;
    },
    {},
  );
  const namedKeys = Object.keys(namedExportsMap);

  const namedExportsCode = `
    var ${namedKey} = ${defaultExport};

    ${namedKeys.map(key => {
      return `var ${namedKey}__${key} = ${namedExportsMap[key]};`
    }).join('\n')}

    export {${namedKeys.map(key => `${namedKey}__${key} as ${key}`).join(',')}};
    
    export default ${namedKey};
  `;
  
  return {
    ...transformResult,
    code: esmCode.replace(defaultExportRegex, namedExportsCode),
  };
}

function removeComments(
  code: string,
  exp: RegExp,
  replaceValue?: string
): string {
  const matches = code.matchAll(exp);
  let matcheStr: string;
  for (let item of matches) {
    matcheStr = item[0];
    if (matcheStr.search(requireRegex) == -1) {
      continue;
    }
    if (!replaceValue) {
      replaceValue = item[1] || '';
    }
    code = code.replace(matcheStr, replaceValue);
  }
  return code;
}

function randomString(length: number): string {
  const code: string =
    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890";
  let result: string = "";
  for (let index = 0; index < length; index++) {
    result += code[Math.floor(Math.random() * code.length)];
  }
  return result;
}

function isString(text: string) {
  try {
    return typeof eval(text) === "string";
  } catch (err) {
    return false;
  }
}