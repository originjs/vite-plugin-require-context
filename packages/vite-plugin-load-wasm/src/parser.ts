import { decode } from '@webassemblyjs/wasm-parser';
import { traverse } from '@webassemblyjs/ast';

export function getImportObjInfo(binary: any) {
  const decoderOpts = {};
  const ast = decode(binary, decoderOpts);

  let map: Map<string, Array<string>> = new Map<string, Array<string>>();
  traverse(ast, {
    ModuleImport(path) {
      const node = path.node;
      map.set(node.module, (map.get(node.module) || []).concat([node.name]));
    },
  });

  return map;
}
