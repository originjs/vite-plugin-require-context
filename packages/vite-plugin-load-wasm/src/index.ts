import { normalizePath, Plugin } from 'vite';
import { createFilter } from '@rollup/pluginutils';
import { readFileSync } from 'fs';
import { getImportObjInfo } from './parser';
import { join, dirname } from 'path';

type PluginOptions = {
  include?: Array<string>;
  exclude?: Array<string>;
};

const DEFAULT_OPTIONS: PluginOptions = {
  include: [],
  exclude: [],
};

// matching like: 'import wasm from "xxx.wasm"'
export const WASM_IMPORT_STATEMENT_REX = /import\s*(\S*)\s*from\s*['"](.*\.wasm).*/;

export default (options: PluginOptions = {}): Plugin => {
  // init options
  options = Object.assign(DEFAULT_OPTIONS, options);

  // normalize path of include and exclude. Then create a filter.
  const include = options.include!.map(item => normalizePath(item));
  const exclude = options.exclude!.map(item => normalizePath(item));
  const filter = createFilter(include, exclude);

  return {
    name: 'vite:load-wasm',
    enforce: 'pre',

    async transform(code: string, id: string) {
      // check file id
      if (!filter(id)) {
        return null;
      }

      // regular matching and check
      const rexRes = WASM_IMPORT_STATEMENT_REX.exec(code);
      if (rexRes === null) {
        return null;
      }
      if (!rexRes[0] || !rexRes[1] || !rexRes[2]) {
        console.log('vite-plugin-load-wasm: regular matching result fail');
        return null;
      }

      // init matchStr, importName, wasmFileName
      const [matchStr, importName, wasmFileName] = [
        rexRes[0],
        rexRes[1],
        rexRes[2],
      ];

      // get importObj
      const info = getImportObjInfo(
        readFileSync(join(dirname(id), wasmFileName)),
      );
      if (info.size > 1) {
        console.log(
          'vite-plugin-load-wasm: import item number should less than 1',
        );
        return null;
      }
      let importFileName;
      for (let name of info.keys()) {
        importFileName = name;
        break;
      }

      // replace string
      let replaceStr;
      if (!!importFileName) {
        replaceStr = `import init from '${wasmFileName}'
let imports = {}
imports['${importFileName}'] = {${info.get(importFileName)!.join(', ')}}
const ${importName} = await init(imports)
`;
      } else {
        replaceStr = `import init from '${wasmFileName}'
const ${importName} = await init()
`;
      }

      // console.log(code.replace(matchStr, replaceStr));
      return code.replace(matchStr, replaceStr);
    },
  };
};
