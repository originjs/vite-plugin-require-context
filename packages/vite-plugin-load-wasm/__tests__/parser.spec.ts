import { readFileSync } from "fs";
import { join } from "path";
import { getImportObjInfo } from "../src/parser";

test('wasm ast parser', () => {
  const binary = readFileSync(join(__dirname, './assets/collisionDetectionRust_bg.wasm'));
  const info = getImportObjInfo(binary);
  expect(info).toMatchSnapshot();
})
