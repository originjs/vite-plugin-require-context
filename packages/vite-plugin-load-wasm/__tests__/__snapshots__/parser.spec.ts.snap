// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wasm ast parser 1`] = `
Map {
  "./collisionDetectionRust_bg.js" => Array [
    "__wbindgen_json_serialize",
  ],
}
`;
